# core

> Personalized tools to add some functionality to SolveIt

In [ ]:
#| default_exp core

In [ ]:
#| hide
from nbdev.showdoc import *

In [ ]:
#| export
from dialoghelper.core import *
from fastcore.tools import in_notebook
if in_notebook:
    import sys; sys.path.append('..')
from solveit_dmtools import dhb, fab, dhp
__all__ = ['dhb', 'fab', 'dhp']


In [ ]:
add_msg(content=dhb.doc, msg_type='note')

'_3f29bd99'

**Backup Chat for SolveIt using dialoghelper and lisette**

Sometimes we may have a problem in SolveIt while Sonnet is down (E300), or maybe we want a different perspective.

This module helps us to leverage any other LLM that is available to LiteLLM by providing our own keys and the model name.

Usage: 
```python
from solveit_dmtools import dhb

# then in another cell
# bc = dhb.c() to search model names
bc = dhb.c("model-name")
bc("Hi")
```


In [ ]:
#| export
bc = dhb.c()

Please try again by using e.g. `bc = dhb.c('model_name')` with a modle name in
azure/eu/gpt-5-2025-08-07
azure/eu/gpt-5-mini-2025-08-07
azure/eu/gpt-5.1
azure/eu/gpt-5.1-chat
azure/eu/gpt-5.1-codex
azure/eu/gpt-5.1-codex-mini
azure/eu/gpt-5-nano-2025-08-07
azure/global/gpt-5.1
azure/global/gpt-5.1-chat
azure/global/gpt-5.1-codex
azure/global/gpt-5.1-codex-mini
azure/gpt-5.1-2025-11-13
azure/gpt-5.1-chat-2025-11-13
azure/gpt-5.1-codex-2025-11-13
azure/gpt-5.1-codex-mini-2025-11-13
azure/gpt-5
azure/gpt-5-2025-08-07
azure/gpt-5-chat
azure/gpt-5-chat-latest
azure/gpt-5-codex
azure/gpt-5-mini
azure/gpt-5-mini-2025-08-07
azure/gpt-5-nano
azure/gpt-5-nano-2025-08-07
azure/gpt-5-pro
azure/gpt-5.1
azure/gpt-5.1-chat
azure/gpt-5.1-codex
azure/gpt-5.1-codex-mini
azure/us/gpt-5-2025-08-07
azure/us/gpt-5-mini-2025-08-07
azure/us/gpt-5-nano-2025-08-07
azure/us/gpt-5.1
azure/us/gpt-5.1-chat
azure/us/gpt-5.1-codex
azure/us/gpt-5.1-codex-mini
gpt-5
gpt-5.1
gpt-5.1-2025-11-13
gpt-5.1-chat-latest
gpt-5-

In [ ]:
#| exports
bc = dhb.c("openrouter/openai/gpt-5-codex")

The following will be automatically commented out when run, then a prompt cell is added after it with input/output from the other LLM.

In [ ]:
#| exports
bc("Hi, can you use tools?")

**Prompt (openrouter/openai/gpt-5-codex):** Hi, can you use tools?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

I do have access to the `read_url` tool by default. Are there other tools or stored variables from the earlier session that you’d like to bring in? If so, could you let me know which ones so you can add them with `bc.add_tools` or `bc.add_vars`?

In [ ]:
# bc("Can you please read https://raw.githubusercontent.com/AnswerDotAI/fhdaisy/refs/heads/main/README.md and give the elevator pitch and some sample code? No need to store the raw content.")

**Prompt (openrouter/openai/gpt-5-codex):** Can you please read https://raw.githubusercontent.com/AnswerDotAI/fhdaisy/refs/heads/main/README.md and give the elevator pitch and some sample code? No need to store the raw content.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

<details class='tool-usage-details'>

```json
{
  "id": "call_uaCBo7Zuop72MztqdQ1aFfjq",
  "call": {
    "function": "read_url",
    "arguments": {
      "url": "https://raw.githubusercontent.com/AnswerDotAI/fhdaisy/refs/heads/main/README.md",
      "as_md": false,
      "extract_section": false,
      "selector": ""
    }
  },
  "result": "# fhdaisy\n\n`fhdaisy` is a Python wrapper for [DaisyUI](https://daisyui.com/) that\nbrings its compone<TRUNCATED>"
}
```

</details>

**Elevator pitch**

fhdaisy lets you build FastHTML apps with DaisyUI’s polished component styles without hand-writing CSS class strings. Every DaisyUI class becomes a Python component, and modifiers use a concise `-primary` style shorthand, so you compose rich, themeable interfaces while staying in Python.

**Sample code**

```python
from fasthtml.common import Div, H2, P
from fasthtml.jupyter import render
from fhdaisy import Card, CardBody, CardTitle, Btn

ui = Card(
    CardBody(
        CardTitle("Welcome to FastHTML + fhdaisy"),
        P("All the DaisyUI styles, none of the class juggling."),
        Div(
            Btn("Primary", cls='-primary'),
            Btn("Outline", cls='-outline'),
            Btn("Success", cls='-success'),
            cls='space-x-2'
        )
    ),
    cls='shadow-xl bg-base-100 max-w-md'
)

render(ui)
```

The snippet renders a DaisyUI card with styled buttons, showing how the title-case components and modifier shorthand keep the markup clean and Pythonic.

In [ ]:
add_msg(content=dhp.doc, msg_type="note")

'_94fe5acd'

**Dialog Helper for Polya's Problem-Solving Method**

This module provides quick access to Polya's four-stage problem-solving process
through interactive prompts.

Each stage has multiple questions/prompts:
- To preview one you can just print it/type its name in a cell and hit Submit
  - e.g. `dhp.act.next` shows you "(prompt) What is next?".
- To execute one you call it by adding () after the name
  - e.g. `dhp.act.next()` will replace the current message cell with a prompt cell having "What is next?" in it
    - It will be automatically executed, you can hit Esc to stop it and/or Enter to edit the prompt

**TYPICAL FLOW:**
- Start with `dhp.u` (`understand`) (even briefly) to clarify your understanding of the problem
- Move to `dhp.p` (`plan`) to develop initial strategies
- Switch between `plan` and `act` (or `a`, or `x`/`execute`) as you develop and test approaches
- Use `dhp.r` (`review`) to gain deeper understanding of your approach and findings
- You might loop back to other steps after `r` (`review`)

If you feel stuck, run `dhp.help()` and it will submit the prompt cell it creates - SolveIt will help you pick a next prompt!

**UNDERSTAND STAGE - Clarify the problem before solving**
- dhp.u.summary() - Creates a prompt cell asking SolveIt to give a **concise summary** of the problem
- dhp.u.info() - Creates a prompt cell asking SolveIt to inventory **known/unknown information**
- dhp.u.similar() - Creates a prompt cell asking SolveIt if it has **seen a similar problem** before
- dhp.u.lateral() - Creates a prompt cell to **explore problem relationships and scope**
- dhp.u.related() - Creates a prompt cell to **identify similar or simpler problems**
- dhp.u.viz() - Creates a prompt cell asking SolveIt to **create a figure or diagram** to represent the problem
- dhp.u.notation() - Creates a prompt cell asking SolveIt to **pick suitable notation** (symbols for quantities/data, states, transitions)
- dhp.u.simplest() - Creates a prompt cell asking SolveIt for the **simplest way to look at** the problem
- dhp.u.simplify() - Creates a prompt cell asking SolveIt to **separate problem parts** (break down complex conditions into simpler ones)

**PLAN STAGE - Develop strategies and approaches**
- dhp.p.chunks() - Creates a prompt cell asking SolveIt to **break down the problem** into smaller sub-problems
- dhp.p.partial() - Creates a prompt cell asking SolveIt if there's a **smaller part or representation** of the problem to solve
- dhp.p.known_approach() - Creates a prompt cell asking SolveIt to **use a known algorithm or library** to solve the problem
- dhp.p.verifiable() - Creates a prompt cell asking SolveIt **how to verify** if the solution is consistent and correct
- dhp.p.backward() - Creates a prompt cell asking SolveIt to **work backward** from the desired result
- dhp.p.aux() - Creates a prompt cell asking SolveIt to **use an auxiliary element** (variable, diagram, or example) to clarify the path
- dhp.p.analogy() - Creates a prompt cell asking SolveIt to **use analogy or similarity** to relate the problem to a known solution
- dhp.p.review() - Creates a prompt cell asking SolveIt to **critique the plan of attack** (be frank and critical)

**ACT STAGE - Execute your plan while monitoring progress**
- dhp.a.all() - Creates a prompt cell asking SolveIt if we **covered all of the data or examples** for this step
- dhp.a.check() - Creates a prompt cell asking SolveIt if **this step seems correct**
- dhp.a.doubt() - Creates a prompt cell asking SolveIt if we're **using the right approach**
- dhp.a.next() - Creates a prompt cell asking SolveIt **what is next**
- dhp.a.other() - Creates a prompt cell asking SolveIt if there's **another way to look at this**
- dhp.a.partial() - Creates a prompt cell asking SolveIt about **intermediate results or milestones** to aim for
- dhp.a.simpler() - Creates a prompt cell asking SolveIt if there was a **simpler way** to do this step
- dhp.a.symmetry() - Creates a prompt cell asking SolveIt about **symmetries or patterns** in the problem to exploit
- dhp.a.valid() - Creates a prompt cell asking SolveIt if **this step was a valid step**

**REVIEW STAGE - Verify results, reflect on process, and extract lessons**
- dhp.r.all() - Creates a prompt cell asking SolveIt if we **covered all of the data or examples** for this problem
- dhp.r.alter() - Creates a prompt cell asking SolveIt for **alternative solutions or approaches** that might be more efficient or effective
- dhp.r.general() - Creates a prompt cell asking SolveIt if we can **generalize the solution** to other similar problems
- dhp.r.grok() - Creates a note cell with the text "**To consider:** Can I understand the solution without having to perform all the steps?"
- dhp.r.learned() - Creates a prompt cell asking SolveIt **what lessons have been learned** from this
- dhp.r.mistakes() - Creates a prompt cell asking SolveIt about **common mistakes** made
- dhp.r.other() - Creates a prompt cell asking SolveIt if we can **derive the result differently**
- dhp.r.principles() - Creates a prompt cell asking SolveIt to **identify underlying principles or patterns** that emerged during the solution process
- dhp.r.sanity() - Creates a prompt cell asking SolveIt if the **result makes sense** and can be verified by substitution or another method
- dhp.r.simpler() - Creates a prompt cell asking SolveIt if we can **derive the result in a simpler way**
- dhp.r.test() - Creates a prompt cell asking SolveIt for **different ways to test** this


The following is just dhp.help()

Please pick an appropriate next-step/prompt from the below:

doc: UNDERSTAND STAGE - Clarify the problem before solving
----
- dhp.u.summary() - Creates a prompt cell asking SolveIt to give a **concise summary** of the problem
- dhp.u.info() - Creates a prompt cell asking SolveIt to inventory **known/unknown information**
- dhp.u.similar() - Creates a prompt cell asking SolveIt if it has **seen a similar problem** before
- dhp.u.lateral() - Creates a prompt cell to **explore problem relationships and scope**
- dhp.u.related() - Creates a prompt cell to **identify similar or simpler problems**
- dhp.u.viz() - Creates a prompt cell asking SolveIt to **create a figure or diagram** to represent the problem
- dhp.u.notation() - Creates a prompt cell asking SolveIt to **pick suitable notation** (symbols for quantities/data, states, transitions)
- dhp.u.simplest() - Creates a prompt cell asking SolveIt for the **simplest way to look at** the problem
- dhp.u.simplify() - Creates a prompt cell asking SolveIt to **separate problem parts** (break down complex conditions into simpler ones)

info: (prompt) "What information do we have? What information do we not have? What might change as we learn more?"
lateral: (prompt) "Can you relate this problem to a more general or a more specific problem?"
notation: (prompt) "Can we pick a suitable notation (e.g. symbols for quantities/data, states, and transitions)?"
related: (prompt) "Can you think of a related problem that we can solve? It could even be a simpler one we could solve first to help understand this one."
similar: (prompt) "Have you seen a similar problem before?"
simplest: (prompt) "What might be the simplest way to look at this problem?"
simplify: (prompt) "Can we separate the various parts of the problem (e.g. break down complex conditions into simpler ones)?"
summary: (prompt) "Could you please give a concise summary of the problem?"
viz: (prompt) "Can we create a figure or diagram to represent the problem?"

doc: PLAN STAGE - Develop strategies and approaches
----
- dhp.p.chunks() - Creates a prompt cell asking SolveIt to **break down the problem** into smaller sub-problems
- dhp.p.partial() - Creates a prompt cell asking SolveIt if there's a **smaller part or representation** of the problem to solve
- dhp.p.known_approach() - Creates a prompt cell asking SolveIt to **use a known algorithm or library** to solve the problem
- dhp.p.verifiable() - Creates a prompt cell asking SolveIt **how to verify** if the solution is consistent and correct
- dhp.p.backward() - Creates a prompt cell asking SolveIt to **work backward** from the desired result
- dhp.p.aux() - Creates a prompt cell asking SolveIt to **use an auxiliary element** (variable, diagram, or example) to clarify the path
- dhp.p.analogy() - Creates a prompt cell asking SolveIt to **use analogy or similarity** to relate the problem to a known solution
- dhp.p.review() - Creates a prompt cell asking SolveIt to **critique the plan of attack** (be frank and critical)

analogy: (prompt) "Can you use analogy or similarity to relate the problem to a known solution?"
aux: (prompt) "Could we use an auxiliary element (e.g., a variable, diagram, or example) to clarify the path?"
backward: (prompt) "Can we work backward from the desired result?"
chunks: (prompt) "Could you please break down the problem into smaller sub-problems?"
known_approach: (prompt) "Could we use a known algorithm or library to solve the problem, or some of it?"
partial: (prompt) "Is there a smaller part or representation of the problem we could solve?"
review: (prompt) "Could you please critique the plan of attack? Be frank, do not be afraid to be critical."
verifiable: (prompt) "How would we verify if our solution is consistent and correct?"

doc: ACT STAGE - Execute your plan while monitoring progress
----
- dhp.a.doubt() - Creates a prompt cell asking SolveIt if we're **using the right approach**
- dhp.a.other() - Creates a prompt cell asking SolveIt if there's **another way to look at this**
- dhp.a.partial() - Creates a prompt cell asking SolveIt about **intermediate results or milestones** to aim for
- dhp.a.symmetry() - Creates a prompt cell asking SolveIt about **symmetries or patterns** in the problem to exploit
- dhp.a.next() - Creates a prompt cell asking SolveIt **what is next**
- dhp.a.valid() - Creates a prompt cell asking SolveIt if **this step was a valid step**
- dhp.a.check() - Creates a prompt cell asking SolveIt if **this step seems correct**
- dhp.a.simpler() - Creates a prompt cell asking SolveIt if there was a **simpler way** to do this step
- dhp.a.all() - Creates a prompt cell asking SolveIt if we **covered all of the data or examples** for this step

all: (prompt) "Did we cover all of the data or examples for this step?"
check: (prompt) "Does this step seem correct?"
doubt: (prompt) "Are we using the right approach?"
next: (prompt) "What is next?"
other: (prompt) "Is there another way to look at this?"
partial: (prompt) "Are there any intermediate results or milestones that we can aim for?"
simpler: (prompt) "Was there a simpler way we could have done this step?"
symmetry: (prompt) "Are there any symmetries or patterns in the problem that we can exploit?"
valid: (prompt) "Does this step seem to have been a valid step?"

doc: REVIEW STAGE - Verify results, reflect on process, and extract lessons
----
- dhp.r.all() - Creates a prompt cell asking SolveIt if we **covered all of the data or examples** for this problem
- dhp.r.sanity() - Creates a prompt cell asking SolveIt if the **result makes sense** and can be verified by substitution or another method
- dhp.r.grok() - Creates a note cell with the text "**To consider:** Can I understand the solution without having to perform all the steps?"
- dhp.r.learned() - Creates a prompt cell asking SolveIt **what lessons have been learned** from this
- dhp.r.general() - Creates a prompt cell asking SolveIt if we can **generalize the solution** to other similar problems
- dhp.r.alter() - Creates a prompt cell asking SolveIt for **alternative solutions or approaches** that might be more efficient or effective
- dhp.r.other() - Creates a prompt cell asking SolveIt if we can **derive the result differently**
- dhp.r.mistakes() - Creates a prompt cell asking SolveIt about **common mistakes** made
- dhp.r.simpler() - Creates a prompt cell asking SolveIt if we can **derive the result in a simpler way**
- dhp.r.principles() - Creates a prompt cell asking SolveIt to **identify underlying principles or patterns** that emerged during the solution process
- dhp.r.test() - Creates a prompt cell asking SolveIt for **different ways to test** this

all: (prompt) "Did we cover all of the data or examples for this problem?"
alter: (prompt) "Can you think of alternative solutions or approaches that might be more efficient or effective?"
general: (prompt) "Can we generalize the solution to other similar problems?"
grok: (note) "**To consider:** Can I understand the solution without having to perform all the steps?"
learned: (prompt) "What lessons have I learned from this?"
mistakes: (prompt) "What were my common mistakes?"
other: (prompt) "Can we derive the result differently?"
principles: (prompt) "Can you identify any underlying principles or patterns that emerged during the solution process?"
sanity: (prompt) "Does the result make sense? Can we verify by substition or another method?"
simpler: (prompt) "Can we derive the result in a simpler way?"
test: (prompt) "What are some different ways we can test this?"

'**Dialog Helper for Polya\'s Problem-Solving Method**\n\nThis module provides quick access to Polya\'s four-stage problem-solving process\nthrough interactive prompts.\n\nEach stage has multiple questions/prompts:\n- To preview one you can just print it/type its name in a cell and hit Submit\n  - e.g. `dhp.act.next` shows you "(prompt) What is next?".\n- To execute one you call it by adding () after the name\n  - e.g. `dhp.act.next()` will replace the current message cell with a prompt cell having "What is next?" in it\n    - It will be automatically executed, you can hit Esc to stop it and/or Enter to edit the prompt\n\n**TYPICAL FLOW:**\n- Start with `dhp.u` (`understand`) (even briefly) to clarify your understanding of the problem\n- Move to `dhp.p` (`plan`) to develop initial strategies\n- Switch between `plan` and `act` (or `a`, or `x`/`execute`) as you develop and test approaches\n- Use `dhp.r` (`review`) to gain deeper understanding of your approach and findings\n- You might loop back to other steps after `r` (`review`)\n\nIf you feel stuck, run `dhp.help()` and it will submit the prompt cell it creates - SolveIt will help you pick a next prompt!\n\n**UNDERSTAND STAGE - Clarify the problem before solving**\n- dhp.u.summary() - Creates a prompt cell asking SolveIt to give a **concise summary** of the problem\n- dhp.u.info() - Creates a prompt cell asking SolveIt to inventory **known/unknown information**\n- dhp.u.similar() - Creates a prompt cell asking SolveIt if it has **seen a similar problem** before\n- dhp.u.lateral() - Creates a prompt cell to **explore problem relationships and scope**\n- dhp.u.related() - Creates a prompt cell to **identify similar or simpler problems**\n- dhp.u.viz() - Creates a prompt cell asking SolveIt to **create a figure or diagram** to represent the problem\n- dhp.u.notation() - Creates a prompt cell asking SolveIt to **pick suitable notation** (symbols for quantities/data, states, transitions)\n- dhp.u.simplest() - Creates a prompt cell asking SolveIt for the **simplest way to look at** the problem\n- dhp.u.simplify() - Creates a prompt cell asking SolveIt to **separate problem parts** (break down complex conditions into simpler ones)\n\n**PLAN STAGE - Develop strategies and approaches**\n- dhp.p.chunks() - Creates a prompt cell asking SolveIt to **break down the problem** into smaller sub-problems\n- dhp.p.partial() - Creates a prompt cell asking SolveIt if there\'s a **smaller part or representation** of the problem to solve\n- dhp.p.known_approach() - Creates a prompt cell asking SolveIt to **use a known algorithm or library** to solve the problem\n- dhp.p.verifiable() - Creates a prompt cell asking SolveIt **how to verify** if the solution is consistent and correct\n- dhp.p.backward() - Creates a prompt cell asking SolveIt to **work backward** from the desired result\n- dhp.p.aux() - Creates a prompt cell asking SolveIt to **use an auxiliary element** (variable, diagram, or example) to clarify the path\n- dhp.p.analogy() - Creates a prompt cell asking SolveIt to **use analogy or similarity** to relate the problem to a known solution\n- dhp.p.review() - Creates a prompt cell asking SolveIt to **critique the plan of attack** (be frank and critical)\n\n**ACT STAGE - Execute your plan while monitoring progress**\n- dhp.a.all() - Creates a prompt cell asking SolveIt if we **covered all of the data or examples** for this step\n- dhp.a.check() - Creates a prompt cell asking SolveIt if **this step seems correct**\n- dhp.a.doubt() - Creates a prompt cell asking SolveIt if we\'re **using the right approach**\n- dhp.a.next() - Creates a prompt cell asking SolveIt **what is next**\n- dhp.a.other() - Creates a prompt cell asking SolveIt if there\'s **another way to look at this**\n- dhp.a.partial() - Creates a prompt cell asking SolveIt about **intermediate results or milestones** to aim for\n- dhp.a.simpler() - Creates a prompt cell asking SolveIt if there was a **simpler way** to do this step\n- dhp.a.symmetry() - Creates a prompt cell asking SolveIt about **symmetries or patterns** in the problem to exploit\n- dhp.a.valid() - Creates a prompt cell asking SolveIt if **this step was a valid step**\n\n**REVIEW STAGE - Verify results, reflect on process, and extract lessons**\n- dhp.r.all() - Creates a prompt cell asking SolveIt if we **covered all of the data or examples** for this problem\n- dhp.r.alter() - Creates a prompt cell asking SolveIt for **alternative solutions or approaches** that might be more efficient or effective\n- dhp.r.general() - Creates a prompt cell asking SolveIt if we can **generalize the solution** to other similar problems\n- dhp.r.grok() - Creates a note cell with the text "**To consider:** Can I understand the solution without having to perform all the steps?"\n- dhp.r.learned() - Creates a prompt cell asking SolveIt **what lessons have been learned** from this\n- dhp.r.mistakes() - Creates a prompt cell asking SolveIt about **common mistakes** made\n- dhp.r.other() - Creates a prompt cell asking SolveIt if we can **derive the result differently**\n- dhp.r.principles() - Creates a prompt cell asking SolveIt to **identify underlying principles or patterns** that emerged during the solution process\n- dhp.r.sanity() - Creates a prompt cell asking SolveIt if the **result makes sense** and can be verified by substitution or another method\n- dhp.r.simpler() - Creates a prompt cell asking SolveIt if we can **derive the result in a simpler way**\n- dhp.r.test() - Creates a prompt cell asking SolveIt for **different ways to test** this\n'


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Based on where we are in the conversation, I'd recommend starting with:

**`dhp.u.summary()`**

Since we've just been introduced to the `dhp` (Dialog Helper for Polya's method) tool and haven't defined a specific problem yet, beginning with a summary prompt would help clarify what problem we're trying to solve. This follows Polya's method by starting in the UNDERSTAND stage before moving to planning or action.

Would you like to run that, or do you have a specific problem in mind that you'd like to work on? If you already have a problem and want to jump to a different stage, let me know!

In [ ]:
#| exports

# Uncomment and submit the line below if you do not already have a copy of fabric in your fabric folder
#!git clone --depth 1 https://github.com/danielmiessler/fabric.git

In [ ]:
add_msg(content=fab.doc, msg_type="note")

'_9b1e86d1'

**fab - Open Source 'fabric' prompts made quickly available in SolveIt**

This module leverages over 200 open source LLM prompts that are available in
Daniel Miesller's 'fabric' project.

If you import as fab, Submit the following to see an overview of all the prompts:
`fab.p`

**HOW TO USE IT**

Most Common Syntax: `prompt="Your Prompt"` in one cell then `fab.p.pattern_name()` in another, where pattern_name is any of the 200+ available fabric patterns.

**MOST IMPORTANT AND USED OPTIONS AND FEATURES**

- **Variable Targeting**: Use `fab.p.pattern_name('variable_name')` to process content from a specific variable instead of the default 'prompt' variable.

- **Pattern Discovery**: Use `fab.p.help()` (an alias for suggest_pattern()) to get suggestions of which pattern to pick for your prompt.

- **Compression Feature**: Use `fab.compress()` after running a pattern to save tokens by marking the previous cell as skipped and compressing the output to a new note.

- **Default Variable**: Most patterns work with a variable called 'prompt' by default, making it easy to process your main content.

**COMMON PATTERNS**

- For Summarizing Content: `fab.p.summarize()`
- For Explaining Code: `fab.p.explain_code()`
- For Analyzing Claims: `fab.p.analyze_claims()`
- For Extracting Wisdom from Text: `fab.p.extract_wisdom()`
- For Creating Quizzes: `fab.p.create_quiz()`


In [ ]:
fab.p.explain_code

Fabric pattern: explain_code - Analyze/explain code, security tool outputs, and configs.

In [ ]:
prompt = read_url("https://raw.githubusercontent.com/shuane/blogtopod/refs/heads/main/blogtopod.py")

The next cell is running fab.p.explain_code()

(From fab.p.explain_code folded below)
# IDENTITY and PURPOSE

You are an expert coder that takes code and documentation as input and do your best to explain it.

Take a deep breath and think step by step about how to best accomplish this goal using the following steps. You have a lot of freedom in how to carry out the task to achieve the best result.

# OUTPUT SECTIONS

- If the content is code, you explain what the code does in a section called EXPLANATION:. 

- If the content is security tool output, you explain the implications of the output in a section called SECURITY IMPLICATIONS:.

- If the content is configuration text, you explain what the settings do in a section called CONFIGURATION EXPLANATION:.

- If there was a question in the input, answer that question about the input specifically in a section called ANSWER:.

# OUTPUT 

- Do not output warnings or notes—just the requested sections.

# INPUT:

INPUT:


$prompt

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

**EXPLANATION:**

This Python script (`blogtopod.py`) converts blog posts into podcast-style audio files using text-to-speech. Here's what it does:

The script fetches a blog post from a URL, extracts the main content, and converts it to audio using OpenAI's TTS API. It processes the text in chunks (since TTS APIs have size limits), generates audio for each chunk, then combines them into a single MP3 file.

**Key components:**

- Uses `contextkit.read_url` to extract article content from URLs
- Leverages `openai` library for text-to-speech conversion (voice: "alloy")
- Splits text into ~4000 character chunks at sentence boundaries to avoid cutting mid-sentence
- Uses `pydub` to concatenate audio segments into one file
- Saves the final audio as `output.mp3`

**Main workflow:**
1. Parse the URL and extract article text
2. Break text into manageable chunks
3. Generate audio for each chunk via OpenAI TTS
4. Combine all audio segments
5. Export as a single MP3 file

The script is designed to run from command line, taking a blog URL as an argument. It requires an OpenAI API key to function.

---

Would you like me to explain any specific part in more detail, or would you like to explore how you might modify or extend this code?

To go back full-circle, we can have a different LLM correct the results...

In [ ]:
#| exports
bc("That's not quite right - can you compare your answer to https://raw.githubusercontent.com/shuane/blogtopod/refs/heads/main/README.md and correct it?")

After comparing the earlier summary with the project’s README, here’s the corrected explanation:

**Corrected explanation**

- The tool is meant to be run through a marimo app (for example via `uvx marimo run --sandbox …`), not just as a plain command-line script. A browser UI opens where you supply the blog URL and output filenames before clicking “Run!”.
- It pulls the article text (using the context-extraction helper), asks Gemini 2.5 Pro to transform that text into a podcast-style script, then sends the script to OpenAI’s text-to-speech service to produce natural audio.
- When finished, it saves both an MP3 and the generated script (`<output>_script.md`) in the working directory.
- You need API keys for both OpenAI and Google Gemini, set via environment variables, and the project expects Python 3.12 (with a note that `pydub` currently doesn’t support 3.13).

Does that align with what you were expecting, or is there another part of the README you’d like to incorporate?

<details>

- id: `gen-1763931705-PBAYraXcm1Lu0qAyyPPG`
- model: `openai/gpt-5-codex`
- finish_reason: `stop`
- usage: `Usage(completion_tokens=613, prompt_tokens=8177, total_tokens=8790, completion_tokens_details=CompletionTokensDetailsWrapper(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=384, rejected_prediction_tokens=None, text_tokens=None, image_tokens=0), prompt_tokens_details=PromptTokensDetailsWrapper(audio_tokens=0, cached_tokens=7040, text_tokens=None, image_tokens=None, video_tokens=0), cost=0.00843125, is_byok=False, cost_details={'upstream_inference_cost': None, 'upstream_inference_prompt_cost': 0.00230125, 'upstream_inference_completions_cost': 0.00613})`

</details>

**Prompt (openrouter/openai/gpt-5-codex):** That's not quite right - can you compare your answer to https://raw.githubusercontent.com/shuane/blogtopod/refs/heads/main/README.md and correct it?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

<details class='tool-usage-details'>

```json
{
  "id": "call_21oEGgErRcjNIVScdPUuUj5g",
  "call": {
    "function": "read_url",
    "arguments": {
      "url": "https://raw.githubusercontent.com/shuane/blogtopod/refs/heads/main/README.md",
      "as_md": false,
      "extract_section": false,
      "selector": ""
    }
  },
  "result": "# blogtopod: Turn blog posts into podcast episodes\n\n*Convert blog posts into podcast episodes using <TRUNCATED>"
}
```

</details>

After comparing the earlier summary with the project’s README, here’s the corrected explanation:

**Corrected explanation**

- The tool is meant to be run through a marimo app (for example via `uvx marimo run --sandbox …`), not just as a plain command-line script. A browser UI opens where you supply the blog URL and output filenames before clicking “Run!”.
- It pulls the article text (using the context-extraction helper), asks Gemini 2.5 Pro to transform that text into a podcast-style script, then sends the script to OpenAI’s text-to-speech service to produce natural audio.
- When finished, it saves both an MP3 and the generated script (`<output>_script.md`) in the working directory.
- You need API keys for both OpenAI and Google Gemini, set via environment variables, and the project expects Python 3.12 (with a note that `pydub` currently doesn’t support 3.13).

Does that align with what you were expecting, or is there another part of the README you’d like to incorporate?

In [ ]:
#| hide
import nbdev; nbdev.nbdev_export()